### DISCUSSION TITLE
getCollection() should always return fresh array

### DISCUSSION THREAD

**Discussion by nzakas** [withastro/roadmap#604] [Proposal]
## Background

In v2.0.0, `getCollection()` was introduced and always returned a fresh array of entry objects. Some time after that, `getCollection()` was modified to [cache](https://github.com/withastro/astro/blob/0daaa9efba6c7caeffcd23ce279f4f54080e9bd6/packages/astro/src/content/runtime.ts#L99) and return the same array of entry objects for each call to the same collection.

## Problem

I recently ran into a problem where one page was modifying the array of entries in a way that another page didn't want, and it took me several days to figure out why this was happening. Because `getCollection()` is always returning the same array, it's possible that two different files will modify that array leading to the results being unexpected. For example, suppose your index page has code like this:

```js
const posts = await getCollection("blog");

// descending order by date
posts.sort((a, b) => b.data.date - a.data.date);
```

So you're sorting the posts in descending order by date because that's the right way to show the data on the index page. However, somewhere else in the site just wants the default ordering, and has code like this:

```js
// will posts be sorted by descending date?
const posts = await getCollection("blog");
```

It's not clear as a user which of these will get called first and second, nor if that order is deterministic. What if a new file is introduced that also modifies the collection in some way?

## Solution

The easiest solution to this problem is for `getCollection()` to *not* return the same array for each call to the same collection. Instead, it would create a clone of the array. The change would be fairly straightforward at [this line](https://github.com/withastro/astro/blob/0daaa9efba6c7caeffcd23ce279f4f54080e9bd6/packages/astro/src/content/runtime.ts#L104):

```diff
    if (typeof filter === 'function') {
      return entries.filter(filter);
    } else {
-      return entries;
+      return entries.concat();
    }
```

This would allow all of the expensive operations involved with retrieving the entries to still be cached, but have a fresh array returned each time to avoid side effects.

Note: The workaround today is to use a filter function such as `getCollection("blog", () => true)`, which will always return a new array each time. This proposed change would allow us to skip the ceremony of providing that extra filter function.

**Comment by natemoo-re** [comment]
Ah, this seems like a very reasonable request! We'll definitely earmark this for Astro 3.0.

**Comment by bholmesdev** [comment]
Sounds like a reasonable change! Always forget how tough JS mutability can be. I'm +1

**Comment by bluwy** [comment]
I've made the fix in https://github.com/withastro/astro/pull/8022 which should go out in the next patch release.

**Reply by nzakas** [reply]
Awesome, thanks!

**Comment by reggi** [comment]
I'm seeing this still be an issue. When running the server on a collection. I can't describe it, or explain it completely. But I was seeing the sort-order of a collection in-server correctly and in-build reversed.

I don't think the reverse was doing anything `(await getCollection('jobs')).reverse()`

I had to do `(await getCollection('jobs', () => true))` to get the right order.

**Reply by bholmesdev** [reply]
Interesting... if you can repro this would definitely be a bug! `toReversed()` is an option if you're on Node v20 as well
